### PR TITLE
Cal/collision

### PR DESCRIPTION
**Pull request by McCallisterRomer**
Added basic collision detection for oriented bounding boxes.

### CODE REVIEW

File: nc/source/input/Input.cpp [hunk 1]
```diff
@@ -1,3 +1,4 @@
+#include "Input.h"
```
Comment (McCallisterRomer) [inline nc/source/input/Input.cpp:1]: I don't think anything actually changed in this file. Not sure what is happening with the diff.

File: nc/source/physics/PhysicsSystem.cpp [hunk 2]
```diff
@@ -1,11 +1,17 @@
...
+#include "CollisionDetection.inl"
 #include "debug/Utils.h"
 #include "Input.h"
 #include "component/Transform.h"
 #include "Window.h"
 #include "graphics/Graphics.h"
 #include "MainCamera.h"
+#include "Ecs.h"
+
+#include "directx/math/DirectXMath.h"
```
Comment (McCallisterRomer) [inline nc/source/physics/PhysicsSystem.cpp]: remove both dx includes

File: nc/source/physics/PhysicsSystem.h [hunk 3]
```diff
@@ -18,6 +19,10 @@ namespace nc::physics
         public:
             PhysicsSystem(graphics::Graphics* graphics);
 
+            void ClearState();
+
```
Comment (McCallisterRomer) [inline nc/source/physics/PhysicsSystem.h]: Lest you want to catch a case, best to delete that white space.

File: project/Components/BoarMover.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
+#include "BoarMover.h"
```
Comment (McCallisterRomer) [inline project/Components/BoarMover.cpp:1]: Ignore this file - getting deleted in next pr.

File: project/Components/BoarMover.h [hunk 5]
```diff
@@ -0,0 +1,21 @@
+#pragma once
```
Comment (McCallisterRomer) [inline project/Components/BoarMover.h:1]: Ignore this file - getting deleted in next pr.

File: nc/include/component/Collider.h [hunk 6]
```diff
@@ -11,15 +11,24 @@ namespace nc
         public:
             Collider(EntityHandle handle, Vector3 scale);
             ~Collider();
-            void Update(graphics::FrameManager& frame);
+
+            void UpdateTransformationMatrix();
+            DirectX::FXMMATRIX GetTransformationMatrix() const;
 
             #ifdef NC_EDITOR_ENABLED
+            void UpdateModel(graphics::FrameManager& frame);
```
Comment (jaremieromer) [inline nc/include/component/Collider.h]: "Model" is a bit confusing as a name here, even though that's what it is. When I see this I think the collider somehow has control over the model it's attached to (Bee, Sphere, etc.) and its not apparent this is referring to the wireframe collision widget.

File: nc/include/component/Collider.h [hunk 7]
```diff
@@ -11,15 +11,24 @@ namespace nc
...
 
-        private:
-            Transform* m_transform;
+        protected:
+            DirectX::XMMATRIX m_colliderMatrix;
+            DirectX::FXMMATRIX m_transformMatrix;
             Vector3 m_scale;
+
+            #ifdef NC_EDITOR_ENABLED
             graphics::Model m_model;
```
Comment (jaremieromer) [inline nc/include/component/Collider.h]: See other comment - I recommend changing this name to m_boundingBox or m_bounds

File: nc/source/component/Collider.cpp [hunk 8]
```diff
@@ -4,38 +4,75 @@
...
-#endif
 
 namespace
 {
-    const auto ColliderMesh = std::string{"project/Models/cube.fbx"};
-    auto GetColliderMaterial = nc::graphics::Material::CreateMaterial<nc::graphics::TechniqueType::Wireframe>;
+    const auto CubeMeshPath = std::string{"project/Models/cube.fbx"};
+    auto CreateMaterial = nc::graphics::Material::CreateMaterial<nc::graphics::TechniqueType::Wireframe>;
+
+    auto CreateBoxModel()
```
Comment (jaremieromer) [inline nc/source/component/Collider.cpp]: See comments about not using "Model" as a name. What about "BoundingBox" or "BoundingSphere" "BoundingCapsule" etc

File: nc/source/component/Collider.cpp [hunk 9]
```diff
@@ -4,38 +4,75 @@
...
+
+        m_model.SetTransformationMatrix(m_colliderMatrix);
+        m_model.Submit(frame);
+    }
+
     void Collider::EditorGuiElement()
     {
         ImGui::Text("Collider");
         ui::editor::xyzWidgetHeader("     ");
         ui::editor::xyzWidget("Scale", "colliderscale", &m_scale.x, &m_scale.y, &m_scale.z, 0.01f, 100.0f);
```
Comment (jaremieromer) [inline nc/source/component/Collider.cpp:70]: Minor nit, but camelCase "colliderscale"

File: nc/source/graphics/d3dresource/ConstantBufferResources.cpp [hunk 10]
```diff
@@ -13,7 +13,7 @@ namespace nc::graphics::d3dresource
     TransformConstBufferVertexPixel::Transforms TransformConstBufferVertexPixel::GetTransforms() noexcept
     {
         const auto gfx = GraphicsResourceManager::GetGraphics();
-        const auto modelView = m_parent.GetTransformXM() * gfx->GetViewMatrix();
+        const auto modelView = m_parent.GetTransformationMatrix() * gfx->GetViewMatrix();
```
Comment (jaremieromer) [inline nc/source/graphics/d3dresource/ConstantBufferResources.cpp:16]: I prefer this naming

File: nc/source/graphics/d3dresource/GraphicsResource.cpp [hunk 11]
```diff
@@ -39,7 +39,8 @@ namespace nc::graphics::d3dresource
 
     Stencil::Stencil(Mode mode)
     : m_tag {stencilTagFromMode(mode)},
-      m_mode {mode}
+      m_mode {mode},
+      m_stencil{nullptr}
```
Comment (jaremieromer) [inline nc/source/graphics/d3dresource/GraphicsResource.cpp:43]: Thanks!

File: nc/source/graphics/Model.cpp [hunk 12]
```diff
@@ -49,12 +49,12 @@ namespace nc::graphics
         return m_indexBuffer->GetCount();
     }
 
-    void Model::UpdateTransformationMatrix(DirectX::FXMMATRIX matrix) noexcept
+    void Model::SetTransformationMatrix(DirectX::FXMMATRIX matrix) noexcept
```
Comment (jaremieromer) [inline nc/source/graphics/Model.cpp:52]: I like that you have changed this name as it is truly a set

File: nc/source/input/Input.cpp [hunk 13]
```diff
@@ -67,42 +68,41 @@ namespace nc::input
         return (GetAsyncKeyState((KeyCode_t)keyCode) & (1 << 15)) == 0 ? false : true;
     }
```
Comment (jaremieromer) [inline nc/source/input/Input.cpp:70]: Looks like spacing was the only change.

File: nc/source/physics/CollisionDetection.inl [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        static std::vector<CollisionData> previousCollisions{}; //can't remain static b/c engine can't clear it on scene change
+        std::vector<CollisionData> currentCollisions;
+        currentCollisions.reserve(previousCollisions.size());
+        BuildCurrentFrameState(colliders, currentCollisions);
+        CompareFrameStates(previousCollisions, currentCollisions);
+        previousCollisions = std::move(currentCollisions);
+    }
+
+    inline void BuildCurrentFrameState(const std::vector<Collider*>& colliders, std::vector<CollisionData>& out)
```
Comment (jaremieromer) [inline nc/source/physics/CollisionDetection.inl:44]: Awesome work!

File: nc/source/physics/CollisionDetection.inl [hunk 15]
```diff
@@ -0,0 +1,119 @@
...
+        auto prevEnd = previous.cend();
+        for(auto& curr : current)
+        {
+            if(prevEnd == std::find(prevBeg, prevEnd, curr))
+                NotifyCollisionEvent(curr, CollisionEvent::Enter);
+        }
+        NC_PROFILE_END();
+    }
+
+    inline void NotifyCollisionEvent(const CollisionData& data, CollisionEvent type)
```
Comment (jaremieromer) [inline nc/source/physics/CollisionDetection.inl:85]: This method especially and this whole file is really easy to read and understand.

File: nc/source/physics/PhysicsSystem.cpp [hunk 16]
```diff
@@ -46,6 +52,16 @@ namespace nc::physics
         impl = this;
     }
 
+    void PhysicsSystem::ClearState()
+    {
+        m_clickableComponents.clear();
+    }
+
+    void PhysicsSystem::DoPhysicsStep(const std::vector<Collider*>& colliders)
```
Comment (jaremieromer) [inline nc/source/physics/PhysicsSystem.cpp:55]: Starting to really be a physics library now!

File: nc/source/physics/PhysicsSystem.h (hunk 3 above)
Comment (jaremieromer) [reply nc/source/physics/PhysicsSystem.h]: lol! Trifle not with unseemly codebases.

File: project/Components/BoarMover.cpp (hunk 4 above)
Comment (jaremieromer) [reply project/Components/BoarMover.cpp:1]: ok

**Review by jaremieromer — APPROVED** [review]
Hooray!!!!!

**Pull request merged**